helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,336 @@
+use helix_core::{coords_at_pos, encoding};
+use helix_view::{
+    document::{Mode, SCRATCH_BUFFER_NAME},
+    graphics::Rect,
+    theme::Style,
+    Document, Editor, Theme, View,
+};
+
+use crate::ui::ProgressSpinners;
+
+use helix_view::editor::StatusLineElement as StatusLineElementID;
+use tui::buffer::Buffer as Surface;
+use tui::text::{Span, Spans};
+
+pub struct RenderContext<'a> {
+    pub doc: &'a Document,
+    pub view: &'a View,
+    pub theme: &'a Theme,
+    pub focused: bool,
+    pub spinners: &'a ProgressSpinners,
+    pub buffer: RenderBuffer<'a>,
+}
+
+impl<'a> RenderContext<'a> {
+    pub fn new(
+        doc: &'a Document,
+        view: &'a View,
+        theme: &'a Theme,
+        focused: bool,
+        spinners: &'a ProgressSpinners,
+    ) -> Self {
+        RenderContext {
+            doc,
+            view,
+            theme,
+            focused,
+            spinners,
+            buffer: RenderBuffer::default(),
```

Comment:
I am a bit confused with the term `buffer` here, maybe `parts`?

Reply:
Yea, on the one hand, this is basically our rendering buffer. But at the same time, it's consisting of `Spans` (`spans`?).
If the name `parts` sounds more intuitive, I can ride with that as well.